pkg/cli/tail.go

```diff
@@ -0,0 +1,257 @@
+			`organizations/[ORGANIZATION_ID], billingAccounts/[BILLING_ACCOUNT_ID]`,
+	})
+
+	f.BoolVar(&cli.BoolVar{
+		Name:    "validate",
+		Aliases: []string{"v"},
+		Target:  &c.flagValidate,
+		Default: false,
+		Usage:   `Turn on for lumberjack log validation`,
+	})
+
+	f.IntVar(&cli.IntVar{
+		Name:    "max-num",
+		Aliases: []string{"n"},
+		Target:  &c.flagMaxNum,
+		Default: 1,
+		Usage:   `Maximum number of most recent logs to validate, default is 1`,
+	})
+
+	f.DurationVar(&cli.DurationVar{
+		Name:    "duration",
+		Aliases: []string{"d"},
+		Target:  &c.flagDuration,
+		Example: "4h",
+		Default: 2 * time.Hour,
+		Usage: `Log filter that determines how far back to search for log ` +
+			`entries, default is 2 hours`,
+	})
+
+	f.StringVar(&cli.StringVar{
+		Name:    "additional-filter",
+		Target:  &c.flagAdditionalFilter,
+		Example: `resource.type = "gae_app" AND severity = ERROR`,
+		Usage: `Log filter in addition to lumberjack log filter used to tail ` +
+			`log entries, see more on ` +
+			`https://cloud.google.com/logging/docs/view/logging-query-language`,
+	})
+
+	f.StringVar(&cli.StringVar{
+		Name:   "override-filter",
+		Target: &c.flagOverrideFilter,
+		Hidden: true,
+		Usage: `Override lumberjack log filter, when it is used, it will be ` +
+			`the only filter used to tail logs`,
+	})
+
+	f.BoolVar(&cli.BoolVar{
+		Name:    "additional-check",
+		Target:  &c.flagAdditionalCheck,
+		Default: false,
+		Usage: `Use it with -validate flag to validate logs tailed with ` +
+			`additional lumberjack specific checks on log labels.`,
+	})
+
+	return set
+}
+
+func (c *TailCommand) Run(ctx context.Context, args []string) error {
+	f := c.Flags()
+	if err := f.Parse(args); err != nil {
+		return fmt.Errorf("failed to parse flags: %w", err)
+	}
+	args = f.Args()
+	if len(args) > 0 {
+		return fmt.Errorf("unexpected arguments: %q", args)
+	}
+
+	if c.flagScope == "" {
+		return fmt.Errorf("scope is required")
+	}
+
+	// Request with negative and greater than 1000 (log count limit) is rejected.
+	if c.flagMaxNum <= 0 || c.flagMaxNum > 1000 {
+		return fmt.Errorf("--max-num must be greater than 0 and less than 1000")
+	}
+
+	// Tail logs.
+	ls, err := c.tail(ctx)
+	if err != nil {
+		return err
+	}
+	if len(ls) == 0 {
+		c.Outf("No logs found.")
+		return nil
+	}
+
+	var extra []validation.Validator
+	if c.flagAdditionalCheck {
+		extra = append(extra, validation.ValidateLabels)
+	}
+
+	// Output results.
+	var retErr error
+	for _, l := range ls {
+		js, err := protojson.Marshal(l)
+		if err != nil {
+			retErr = errors.Join(retErr, fmt.Errorf("failed to marshal log to json (InsertId: %q): %w", l.InsertId, err))
+			continue
+		}
+
+		// Output tailed log.
+		c.Outf(string(js))
+
+		// Output validation result if validation is enabled.
+		if c.flagValidate {
+			if err := validation.Validate(string(js), extra...); err != nil {
+				retErr = errors.Join(retErr, fmt.Errorf("failed to validate log (InsertId: %q): %w", l.InsertId, err))
```

Comment:
We should output error right away `c.Errf` so that the error message will be right after the log itself. Otherwise it would be hard to map the error message back to the log.